### ISSUE TITLE
[Elastic Agent] Provide additional metadata in the agent local_metadata

### ISSUE THREAD

**Issue by nimarezainia** [elastic/elastic-agent#3213]
**Describe the enhancement:**
Elastic agent's local_metadata to be enhanced to include cloud metadata information. The [cloud metadata processor](https://www.elastic.co/guide/en/beats/filebeat/current/add-cloud-metadata.html) provides exactly this functionality. By enhancing the local_metadata users in Fleet will have greater power to search for agents matching a certain cloud attribute.
Further more, when we look at enhancing the way Fleet displays agent data at scale, this cloud metadata can be used to slice and dice the list of agents in different graphical way.

Fields should be added to local_metadata according to the cloud provider the agent is deployed on.

**Describe a specific use case for the enhancement or feature:**
This will allow our users to search and manage their agents by cloud metadata and instance tags instead of hostnames and OS metadata.

**What is the definition of done?**

**Comment by pierrehilbert** [comment]
Thx @nimarezainia for the issue creation.
Some points to clarify from my perspective:

1. Should it be something optional (behind a feature flag for example) or always there?
2. What should we do when not on cloud environment?
3. What should be the behavior if we can't get the related information?
4. Should we add every information we have in the cloud metadata processor or part of it?

In addition, there is some work to plan on Fleet UI side to benefit from this so we should probably already create an issue there too.

**Comment by amitkanfer** [comment]
> Should it be something optional (behind a feature flag for example) or always there?

Keep in mind we have the [data fidelity](https://github.com/elastic/ingest-dev/issues/1660) feature, which can come useful here.

**Comment by cmacknz** [comment]
> In addition, there is some work to plan on Fleet UI side to benefit from this so we should probably already create an issue there too.

Likely we would implement this as a new Cloud provider: https://www.elastic.co/guide/en/fleet/current/providers.html

We don't expose provider configuration in Fleet at all, so without UI work the only thing we can do is enable this by default everywhere. The cloud_metadata_processor is enabled by default in all the Beats and can't be changed either until we have global processor configuration, but we do get requests to be able to turn it off.

**Comment by nimarezainia** [comment]
@pierrehilbert 

> 1. Should it be something optional (behind a feature flag for example) or always there?
> 

No.

>     2. What should we do when not on cloud environment?
>

What ever the add_cloud_metadata provides. Probably will be null for those fields,
 
>     3. What should be the behavior if we can't get the related information?
> 

Field is blank. I believe that is what the add_cloud_metadata does.

>     4. Should we add every information we have in the cloud metadata processor or part of it?
> 

Add everything. There are only a few fields.

> 
> In addition, there is some work to plan on Fleet UI side to benefit from this so we should probably already create an issue there too.

No. there shouldn't be any Fleet UI work. The request is to enhance he local_metadata so that the KQL query can pick it up. At the moment this issue is not asking for any UI enhancements.

**Comment by nimarezainia** [comment]
sorry just realized that the description was truncated. fixed now.

**Comment by nimarezainia** [comment]
> > In addition, there is some work to plan on Fleet UI side to benefit from this so we should probably already create an issue there too.
> 
> Likely we would implement this as a new Cloud provider: https://www.elastic.co/guide/en/fleet/current/providers.html
> 
> We don't expose provider configuration in Fleet at all, so without UI work the only thing we can do is enable this by default everywhere. The cloud_metadata_processor is enabled by default in all the Beats and can't be changed either until we have global processor configuration, but we do get requests to be able to turn it off.

@cmacknz the Cloud Provider could work if all the information provided by the processor is included in those key value pairs.

I am not fully grasping the second comment. Could we not add fields to the local_metadata that is searchable in Fleet?
![image](https://github.com/elastic/elastic-agent/assets/13442829/ff87babd-005c-4ef1-a292-905153e55a10)

... by using the function call add_cloud_metadata to get the relevant data from the local machine?
(i'm not sure where Global processors fit in this)

**Comment by jlind23** [comment]
@nimarezainia Please also consider adding the user/group used for running Agent. This will help us work on the unprivileged Agent UX in Fleet.

**Comment by nimarezainia** [comment]
@cmacknz do you think this issue can be completed in a single sprint?

**Comment by cmacknz** [comment]
> @cmacknz do you think this issue can be completed in a single sprint?

I don't think so, I think we may actually need a small RFC to decide how we want to maintain and test this, because today agent has no dependency on a particular cloud provider, and doing this will introduce that. Ideally we just reuse the guts of the `add_cloud_metadata` processor in Beats including all the tests for it, but that would introduce a dependency on the Beats repository which we have been trying to avoid.

> We don't expose provider configuration in Fleet at all, so without UI work the only thing we can do is enable this by default everywhere. The cloud_metadata_processor is enabled by default in all the Beats and can't be changed either until we have global processor configuration, but we do get requests to be able to turn it off.

> @cmacknz the Cloud Provider could work if all the information provided by the processor is included in those key value pairs.

What I meant but this is there will be no way for users to turn off the agent's attempts to fetch cloud metadata. This is a problem today with the `add_cloud_metadata` global processor in Beats, agents that run outside the log are full of pointless error logs trying to determine if they are on AWS. In that sense this is making an existing problem worse, but I do think we'll get complaints about this.

**Comment by nimarezainia** [comment]
thanks @cmacknz 
@pierrehilbert perhaps we can use p26 for the definition and perhaps the enhancement to add agent privilege information to local_metadata. We can then pursue implementation of cloud_metadata at a follow on sprint.

**Comment by jlind23** [comment]
@cmacknz @pierrehilbert can we try to work on this a bit earlier than planned to ship the `unpriviledged`? What we want to send is a boolean saying whether or not we are running in a privileged mode.
How complex do you think this is? Doing only this part will help us unblock the fleet work.

**Comment by pierrehilbert** [comment]
From my perspective, this one is covering only the cloud metadata.
We should follow https://github.com/elastic/elastic-agent/issues/3906 for the unprivileged thing.

**Comment by amitkanfer** [comment]
As this issue block a P1 issue: https://github.com/elastic/ingest-dev/issues/2916, i believe this needs to be P1 as well (vs. P2 it has now). @nimarezainia @jlind23 ?

**Comment by jlind23** [comment]
@amitkanfer according to Pierre's comment we should follow up in https://github.com/elastic/elastic-agent/issues/3906

**Comment by nimarezainia** [comment]
@pierrehilbert I marked https://github.com/elastic/elastic-agent/issues/3906 as P1

**Comment by nimarezainia** [comment]
Linking this relevant issue here: https://github.com/elastic/beats/issues/13816
